client/src/unifyfs.c

```diff
@@ -973,14 +961,664 @@ int unifyfs_fid_create_directory(const char* path)
+            continue;
+        }
+
+        /* if we read all of the bytes, we're done */
+        if (req->nread == req->length) {
+            continue;
+        }
+
+        /* otherwise, we have a short read, check whether there
+         * would be a hole after us, in which case we fill the
+         * request buffer with zeros */
+
+        /* get file size for this file */
+        off_t filesize_offt = unifyfs_gfid_filesize(req->gfid);
+        if (filesize_offt == (off_t)-1) {
+            /* failed to get file size */
+            req->errcode = ENOENT;
+            continue;
+        }
+        size_t filesize = (size_t)filesize_offt;
+
+        /* get offset of where hole starts */
+        size_t gap_start = req->offset + req->nread;
+
+        /* get last offset of the read request */
+        size_t req_end = req->offset + req->length;
+
+        /* if file size is larger than last offset we wrote to in
+         * read request, then there is a hole we can fill */
+        if (filesize > gap_start) {
+            /* assume we can fill the full request with zero */
+            size_t gap_length = req_end - gap_start;
+            if (req_end > filesize) {
+                /* request is trying to read past end of file,
+                 * so only fill zeros up to end of file */
+                gap_length = filesize - gap_start;
+            }
+
+            /* copy zeros into request buffer */
+            char* req_ptr = req->buf + req->nread;
+            memset(req_ptr, 0, gap_length);
+
+            /* update number of bytes read */
+            req->nread += gap_length;
+        }
+    }
+
+    /* if we attempted to service requests from our local extent map,
+     * then we need to copy the resulting read requests from the local
+     * and server arrays back into the user's original array */
+    if (unifyfs_local_extents) {
+        /* TODO: would be nice to copy these back into the same order
+         * in which we received them. */
+
+        /* copy locally completed requests back into user's array */
+        int local_count = in_count - count;
+        if (local_count > 0) {
+            memcpy(in_reqs, local_reqs, local_count * sizeof(read_req_t));
+        }
+
+        /* copy sever completed requests back into user's array */
+        if (count > 0) {
+            /* skip past any items we copied in from the local requests */
+            read_req_t* in_ptr = in_reqs + local_count;
+            memcpy(in_ptr, read_reqs, count * sizeof(read_req_t));
+        }
+
+        /* free storage we used for copies of requests */
+        if (reqs != NULL) {
+            free(reqs);
+            reqs = NULL;
+        }
+    }
+
+    return rc;
+}
+
 /* Write count bytes from buf into file starting at offset pos.
  *
  * Returns UNIFYFS_SUCCESS, or an error code
  */
-int unifyfs_fid_write(int fid, off_t pos, const void* buf, size_t count)
+int unifyfs_fid_write(
+    int fid,          /* global file id to write to */
```

Comment:
again, local file id, not global file id

Reply:
Thanks